File: polyglot-scala/src/main/scala/org/sonatype/maven/polyglot/scala/model/Parent.scala
/**
 * Copyright (c) 2012 to original author or authors
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 */
package org.sonatype.maven.polyglot.scala.model

class Parent(
    val gav: Option[Gav],
    val relativePath: String
)

object Parent {
  def apply(
      gav: Gav = null,
      relativePath: String = "../pom.xml"
  ): Parent =
    new Parent(
      Option(gav),
      relativePath
    )
}

import org.sonatype.maven.polyglot.scala.ScalaPrettyPrinter._

class PrettiedParent(p: Parent) {
  def asDoc: Doc = {
    val args = scala.collection.mutable.ListBuffer[Doc]()
    p.gav.foreach(gav => args += assign("gav", gav.asDoc))
    Some(p.relativePath).filterNot(_ == "../pom.xml").foreach(args += assignString(
      "relativePath",
      _
    ))
    `object`("Parent", args.toList)
  }
}

import org.sonatype.maven.polyglot.scala.MavenConverters._
import org.apache.maven.model.{Parent => MavenParent}

class ConvertibleMavenParent(mp: MavenParent) {
  def asScala: Parent = {
    Parent(
      Option(mp.getArtifactId).map(artifactId =>
        (mp.getGroupId, artifactId, mp.getVersion).asScala
      ).orNull,
      mp.getRelativePath
    )
  }
}

class ConvertibleScalaParent(p: Parent) {
  def asJava: MavenParent = {
    val mp = new MavenParent
    p.gav.foreach {
      gav =>
        mp.setGroupId(gav.groupId.orNull)
        mp.setArtifactId(gav.artifactId)
        mp.setVersion(gav.version.orNull)
    }
    mp.setRelativePath(p.relativePath)
    mp
  }
}

File: polyglot-kotlin/src/main/kotlin/org/sonatype/maven/polyglot/kotlin/dsl/Parent.kt
package org.sonatype.maven.polyglot.kotlin.dsl

@PomDsl
class Parent : org.apache.maven.model.Parent(), Cloneable {

  init {
    relativePath = "../pom.kts"
  }

  @PomDsl
  fun groupId(groupId: String): Parent {
    this.groupId = groupId
    return this
  }

  @PomDsl
  fun artifactId(artifactId: String): Parent {
    this.artifactId = artifactId
    return this
  }

  @PomDsl
  fun version(version: String): Parent {
    this.version = version
    return this
  }

  @PomDsl
  infix fun relativePath(relativePath: String): Parent {
    this.relativePath = relativePath
    return this
  }

  override fun clone(): org.apache.maven.model.Parent {
    return super<org.apache.maven.model.Parent>.clone()
  }
}
